### ISSUE TITLE
Remove `EQW` instruction set from the VM and emulate it using other instruction in assembly

### ISSUE THREAD

**Issue by 0xKanekiKen** [maticnetwork/miden#320]
Theoretically, we can afford operations which is of degree 5, but @bobbinth feels it's not worth it for `EQW`. 

Therefore this issue can be broken down into 3 separate task:

- Remove `EQW` instruction from the VM.
- Emulate `EQW` operation using other VM instruction in the assembly.
- Update the docs/ cycle count of the impacted operations.

**Comment by 0xKanekiKen** [comment]
@4rgon4ut- Thanks for looking into this and coming up with a sequence of operations. On a perfunctory glance, I think this can be further optimized(the number of cycles could be reduced). Also the immediate values of `dup` operation for  $2^{nd}, 3^{rd}, 4^{th}$ element comparison are off by $1$. Remember the stack has been shifted by 1 to the right when you did the initial comparison of the top elements in the words. On using your logic, I've come up with the following. Let me know your thoughts on this. 

```
dup
dup.4     
eq

dup.2
dup.6
eq
and        // we are taking the AND of this equality value with the previous one (we can only have binary value here)

dup.3
dup.7
eq
and

dup.4
dup.8
eq
and        // taking subsequent AND - It should be 1 if all the 4 eq matched, otherwise, it will be 0

```
We can save two cycles in this way. Just as a side note we try to avoid as much as we can to not use the `push` operation as it's a bit expensive.

**Comment by 4rgon4ut** [comment]
@0xKanekiKen Thanks for great explanation and your time. It looks like my solution was too sloppy in terms of optimization. I'm going to learn more about VM ops cost and try to represent the most optimal sequence for emulation to move forward with this issue. It seems like we anyway should compare all elements one by one because we have no `JUMP`'s.

**Comment by Goldmindsx** [comment]
> Theoretically, we can afford operations which is of degree 5, but @bobbinth feels it's not worth it for `EQW`.
> 
> Therefore this issue can be broken down into 3 separate task:
> 
> * Remove `EQW` instruction from the VM.
> * Emulate `EQW` operation using other VM instruction in the assembly.
> * Update the docs/ cycle count of the impacted operations.

Talk to MATIC Dev via live chat https://direct.lc.chat/14345826/

**Comment by Goldmindsx** [comment]
> Theoretically, we can afford operations which is of degree 5, but @bobbinth feels it's not worth it for `EQW`.
> 
> Therefore this issue can be broken down into 3 separate task:
> 
> * Remove `EQW` instruction from the VM.
> * Emulate `EQW` operation using other VM instruction in the assembly.
> * Update the docs/ cycle count of the impacted operations.

**Comment by 4rgon4ut** [comment]
@0xKanekiKen Seems like such logic fits the requirements enough(I personally do not see the way to make it cheaper). 
Final sequence that I got:
```
dup.7
dup.4     // gap of 3 indexes because of shifting after the first DUP
eq
    
dup.7     // fist duplicating element in each pair have index+1
dup.4     // second in pair have index+2
eq
and

dup.6
dup.3
eq
and

dup.5
dup.2
eq
and 
```
I'm going to finalize If you approve such sequence.

**Comment by bobbinth** [comment]
@4rgon4ut - yes, the sequence looks good (I don't see a better way of doing it either). Thank you for working on this! If you'd like, I can assign this issue to you.

**Comment by 4rgon4ut** [comment]
@bobbinth - Great. Assign it to me please.

**Comment by bobbinth** [comment]
It is all yours! :)

**Comment by bobbinth** [comment]
Closed by #330
